packages/terra-folder-tree/src/FolderTree.jsx

```diff
@@ -128,6 +139,20 @@ const FolderTree = ({
     listItems[0].tabIndex = 0;
   }, []);
 
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  useEffect(() => {
+    const listItems = listNode.current.querySelectorAll('[data-item-show-focus=true]');
+    const visibleListItems = getVisibleListItems(listItems);
+
+    if (visibleListItems.length === children.length) {
+      setAriaLiveMessage(intl.formatMessage({ id: 'Terra.folder-tree.button.collapse-all-announcement' }));
+    }
+
+    if (visibleListItems.length === listItems.length) {
+      setAriaLiveMessage(intl.formatMessage({ id: 'Terra.folder-tree.button.expand-all-announcement' }));
+    }
+  });
```

Comment:
I was wondering if making those calculations on every re-render can become a problem in case of a big folder tree? I was thinking that moving the functionality implemented in this useEffect to a method, then passing a callback to that method to `FolderItem` to call on collaps/expand events might improve the performance.

Reply:
I was trying to think of a way to only get the new `visibleListItems` if `onExpandAll` or `onCollapseAll` had been called. The problem is that `visibleListItems` isn't changed until after the component renders, so we can't simply make a check on the expand all/collapse all events.

> I was thinking that moving the functionality implemented in this useEffect to a method, then passing a callback to that method to FolderItem to call on collaps/expand events might improve the performance.

The full folder tree expand/collapse does not trigger the row-level expand collapse so unfortunately this doesn't work.

Reply:
This is the best solution I've found so far to prevent the useEffect from running on every render: [c093e25](https://github.com/cerner/terra-framework/pull/2005/commits/c093e25686488d26a43dd4b2d16f53a6f0d1f7a2)